src/System.CodeDom/src/System/CodeDom/Compiler/TempFiles.cs

```diff
@@ -123,6 +122,10 @@ private void EnsureTempNameCreated()
                     catch (IOException)
                     {
                         retryCount--;
+                        if (retryCount == 0)
```

Comment:
Given how you found this, I think we actually need to put back the second half of the condition as well.  If a non-existent directory is provided, we don't want to try 5000 different files, all of which will fail, before actually failing: we want to fail immediately.  But we also can't put it back exactly as it was, as it won't work cross-platform, due to different .  I suggest this:
```C#
if (retryCount == 0 || e is DirectoryNotFoundException)
{
    throw;
}
```
It's not 100% the same, but it addresses the most likely case (unfortunately there's no strongly-typed "file already exists" exception we can special case).  If you can think of other exceptions that we'd get here, we can add those, too.